### PR TITLE
kv: stop concerning the TxnCoordSender with abandoned txns

### PR DESCRIPTION
**Pull request by andreimatei**
Assorted cleanups of the client.Txn interface and the TxnCoordSender. See individual commits. The main point is the last one, which reads:

kv: stop concerning the TxnCoordSender with abandoned txns  …
Before this patch, the TxnCoordSender had a role in cleaning up
abandoned transactions, where abandoned was defined in a funky way:
1) for txns with a cancelable context, a txn was abandoned if the ctx
had been canceled.
2) for the other txns, the txn was considered abandoned if it ran for 10
seconds.

SQL only triggers 1).

The 10 second timeout was a relicv of the past, from a time where the
coordinator was possibly removed from the client. No more, they have
been collocated for years.
In fact, the whole idea of the TCS detecting abandoned txns is a relicv
of that past: all the txn users are already always cleaning up after
themselves.

The pressing motivation for this change is that there's an issue with
checking the context: there is currently no way to pass a ctx with the
same lifetime as the txn to the TxnCoordSender (as it is only a Sender).
And so the TCS was capturing the ctx of the first Send() op as the ctx
of the txn: this was a major hack that's a big problem - we can't have
per-statement contexts, for example. This also restricts the possible
implementations of statement cancelation: we have been forced to cancel
the whole transaction, which is not ideal.

This patch removes TCS' involvement in cleanup by relying on the client
to always send an EndTransaction. SQL was already doing that, as was
everybody else. The TCS is simplified.

The GCQueue maintains its role in detecting abandoned transactions - it
coninues to do so with an one hour timeout. This detection is needed, of
course, since the client is not collocated with the txn's range.

This patch removes the txn.Abandons metric which was maintained by the
TCS. There is a similar metric maintained by the GCQueue. The TCS one
was incorrect anyway, as it was being incremented on ctx cancelation.

Release note: None

### CODE REVIEW

**Comment by andreimatei** [comment]
Spencer, I can find another reviewer if you don't want to do it.
cc @jordanlewis to note that the nasty ctx capture in TxnCS is going away.

There's a commented out test for which I have to figure out what to do: it's spirit might be worth preserving.

**Comment by bdarnell** [comment]
:lgtm: for 2.1 (but I wouldn't cherry-pick it to 2.0, and so we might want to wait to merge this to minimize complications from any other bug-fixing that may go on in this area in the next few weeks)

---

Reviewed 1 of 1 files at r2, 1 of 1 files at r4, 2 of 2 files at r5, 1 of 2 files at r6, 2 of 2 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender_test.go, line 1368 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/23055#-L6Dvwy-_EqS4DeiWnI-:-L6Dvwy-_EqS4DeiWnI0:br0bnkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf7177869cf7fcf965d03a0c3d37fd31abc367f/pkg/kv/txn_coord_sender_test.go#L1368)):*
> ```Go
> 
> // !!!
> // // TestTxnReadAfterAbandon checks the fix for the condition in issue #4787:
> ```

This seems fine to delete to me - how would this test make sense in the absence of abandonment?

---

*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx, line 40 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/23055#-L6DvRkcq4Q9TUyLMn19:-L6DvRkdEyHxvtbnsBLi:bzfvt8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf7177869cf7fcf965d03a0c3d37fd31abc367f/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx#L40)):*
> ```XML
>         <Metric name="cr.node.txn.commits1PC" title="Fast-path Committed" nonNegativeRate />
>         <Metric name="cr.node.txn.aborts" title="Aborted" nonNegativeRate />
>         // !!!
> ```

I'd just remove the next line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23055#-:-L6DwVlf2jkb6zcRG9G0:b2z7rr)*

**Comment by andreimatei** [comment]
Yeah, I definitely won't cherry-pick this, it's the kind of change that can have some unintended consequences. For example I already know one that I need to figure out a story for - we sometimes cancel high-level contexts that still result in sending rollback requests (for example when the network connection drops). Except the rollbacks never get sent because the context is canceled (and so gRPC or someone else along the way refuses to send anything). I believe things used to work to some extent because the heartbeat loop would have detected the cancelation and send an `EndTransaction` using a different, not canceled, context specially built for the purpose. So we used to both call something like `txn.CleanupOnError()` - which would have surprisingly been a no-op, and independently the heartbeat loop would send an `EndTransaction` that actually does something. I need to figure out what the right thing to do is to make sure that we can send rollbacks even with a canceled ctx.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender_test.go, line 1368 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/23055#-L6Dvwy-_EqS4DeiWnI-:-L6DxqRShYXbQROlXCuL:b-j7x422) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf7177869cf7fcf965d03a0c3d37fd31abc367f/pkg/kv/txn_coord_sender_test.go#L1368)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems fine to delete to me - how would this test make sense in the absence of abandonment?
</blockquote></details>

well, what I believe makes sense to test is that sending through a txn whose heartbeat had previously failed (async) is somehow rejected

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23055#-:-L6Dy9IhyBDreExgBQgw:b-2xedex)*

File: pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx [hunk 1]
```diff
@@ -37,6 +37,7 @@ export default function (props: GraphDashboardProps) {
         <Metric name="cr.node.txn.commits" title="Committed" nonNegativeRate />
         <Metric name="cr.node.txn.commits1PC" title="Fast-path Committed" nonNegativeRate />
         <Metric name="cr.node.txn.aborts" title="Aborted" nonNegativeRate />
+        // !!!
```
Comment (couchand) [inline pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx]: ???

**Comment by andreimatei** [comment]
note to self: plan is to check ctx status in `txn.rollback()` and send a (possibly second) `EndTxn` on a different ctx.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LBXC3Gi800hlLBSlzro:bqc4kwo)*

**Comment by andreimatei** [comment]
Ben, please take another look.
I've made `txn.Rollback()` work with a canceled context.

There's 2 things left - I've made `TestHeartbeatCallbackForDecommissioning` flaky. I hope @tschottdorf can tell me whats up.
And I've added a `TestCleanupWithCanceledContext` to check the `txn.Rollback()` behavior. Except it passes with the old code too. I think it only fails if the txn record is on a remote node - so I need an RPC. And so I need a split in the keys so I can do my txn on a remote range. Any suggestion for the easiest/best way to write that test would be most welcomed.

---

Review status: 0 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx, line 40 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-L6DvRkcq4Q9TUyLMn19:-LBhnMkc11LHsH1oHdRz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf7177869cf7fcf965d03a0c3d37fd31abc367f/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx#L40)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd just remove the next line.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx, line 40 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-L69fwZKinOnzwZu61tD-r8-40:-LBhnMzA8h1ls_1syxO1:b8zfe59) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebf7177869cf7fcf965d03a0c3d37fd31abc367f/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/distributed.tsx#L40)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

???
</blockquote></details>

I was asking about just removing this, but not in so many words.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LBhuiDc2pbZWoApMtCl:b-nfrd9q)*

**Comment by bdarnell** [comment]
Reviewed 1 of 1 files at r10, 1 of 2 files at r13, 1 of 1 files at r15, 18 of 18 files at r16.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/internal/client/txn.go, line 106 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBrZ8dW286lGYArRAL-:-LBrZ8dXCzajCJ5HJGfm:bf1mtoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L106)):*
> ```Go
> // distributed transactions (leaf).
> //
> // If the transactions is used to send any operations, CommitOrCleanup() or
> ```

s/transactions/transaction/

---

*[pkg/internal/client/txn.go, line 599 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBrZLgTAhppM3wtqkDX:-LBrZLgTAhppM3wtqkDY:b-fefcdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L599)):*
> ```Go
> }
> 
> func (txn *Txn) rollback(ctx context.Context) *roachpb.Error {
> ```

I assume this copy of the method is going away?

---

*[pkg/internal/client/txn.go, line 631 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBrZv7R58O_jeMmbnFo:-LBrZv7R58O_jeMmbnFp:bny6fj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L631)):*
> ```Go
> 
> 	ctx = txn.db.AnnotateCtx(context.Background())
> 	go func() {
> ```

I think this should be a stopper task. 

---

*[pkg/internal/client/txn.go, line 687 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBr_9cw1s1dL5ulOzMi:-LBr_9cw1s1dL5ulOzMj:b-uigi5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/02d42700c8eaeaca36ce9191a4bea93ee5ee9cf3/pkg/internal/client/txn.go#L687)):*
> ```Go
> // TxnExecOptions controls how Exec() runs a transaction and the corresponding
> // closure.
> type TxnExecOptions struct {
> ```

I'm glad to see this simplification.

---

*[pkg/internal/client/txn.go, line 875 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBr_EGZCMyYD2UBbILI:-LBr_EGZCMyYD2UBbILJ:bmipkj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L875)):*
> ```Go
> 		sender = txn.mu.sender
> 		if txn.mu.Proto.Status != roachpb.PENDING || txn.mu.finalized {
> 			onlyRolback := lastIndex == 0 && haveEndTxn && !endTxnRequest.Commit
> ```

s/Rolback/Rollback/

---

*[pkg/kv/txn_coord_sender.go, line 960 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBr_zUTE_uQyTIzEEEM:-LBr_zUTE_uQyTIzEEEN:bltwdw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/kv/txn_coord_sender.go#L960)):*
> ```Go
> 			return
> 		case <-tc.stopper.ShouldQuiesce():
> 			// TODO(andrei): should we tryAsyncAbort() here?
> ```

Probably not. This is the point at which new tasks can't be created so I'm not sure if it would even work. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055)*

**Comment by andreimatei** [comment]
first 3 commits are #25541

I've added a commit getting rid of the TCS async aborts.
PTAL, a bunch of stuff changed in the `kv: stop concerning the TxnCoordSender with abandoned txns` commit too.
Thanks!
cc @tschottdorf 

Note to self: DNM until #25586 is elucidated.

---

Review status: 0 of 39 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/internal/client/txn.go, line 106 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBrZ8dW286lGYArRAL-:-LCfWg3F57q3NE5sJelY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L106)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/transactions/transaction/
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 599 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBrZLgTAhppM3wtqkDX:-LCfWJ5R9BTJNpvPgJ9o:b-5g7r8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L599)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I assume this copy of the method is going away?
</blockquote></details>

yes, gone

---

*[pkg/internal/client/txn.go, line 631 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBrZv7R58O_jeMmbnFo:-LCfXDRV427Tk05TA99-:b-d7bhn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L631)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think this should be a stopper task. 
</blockquote></details>

will do

---

*[pkg/internal/client/txn.go, line 875 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBr_EGZCMyYD2UBbILI:-LCfWatA4PaXsywN__EA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/internal/client/txn.go#L875)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/Rolback/Rollback/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LCfXfHl6jS9hjJ-kzyA:b-ihbnrz)*

**Comment by andreimatei** [comment]
please give this bad boy another look, it has evolved some

---

Review status: 0 of 22 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/internal/client/txn.go, line 687 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBr_9cw1s1dL5ulOzMi:-LDxGyDVEcaY35tJ3_Rp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02d42700c8eaeaca36ce9191a4bea93ee5ee9cf3/pkg/internal/client/txn.go#L687)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm glad to see this simplification.
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 960 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LBr_zUTE_uQyTIzEEEM:-LDxGwnA5dOLZ1r8R3B7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b413daf0e249a3c285ea1bef151ea595f35cdf99/pkg/kv/txn_coord_sender.go#L960)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Probably not. This is the point at which new tasks can't be created so I'm not sure if it would even work. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LEFbxoj16uODHYnXQ95:b-vn1kt7)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 1252 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEFfJ2QBx_uvkBLbf-8:-LEFfJ2QBx_uvkBLbf-9:b-fzccps) ([raw file](https://github.com/cockroachdb/cockroach/blob/c227e393d7ebb044fc6caa13a1d3ac025fbd6d23/pkg/kv/txn_coord_sender.go#L1252)):*
> ```Go
> 			log.VEventf(ctx, 2, "resetting epoch-based coordinator state on retry")
> 			tc.mu.Lock()
> 			tc.mu.meta.CommandCount = 0
> ```

Is this a duplicate of some existing code or is reviewable just confusing me? We shouldn't copy this list of variables that need to be reset in multiple places.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LEFfog50bLrS0ltqlsL:bb6ywmx)*

**Comment by andreimatei** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 1252 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEFfJ2QBx_uvkBLbf-8:-LEMj4wbEIN-6RToibW5:b-4xx2e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c227e393d7ebb044fc6caa13a1d3ac025fbd6d23/pkg/kv/txn_coord_sender.go#L1252)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this a duplicate of some existing code or is reviewable just confusing me? We shouldn't copy this list of variables that need to be reset in multiple places.
</blockquote></details>

this list just moved down from a few lines above. It belongs better after the `if errTxnID != newTxn.ID {` check above

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055)*

**Comment by nvanbenschoten** [comment]
:lgtm: although `s/relicv/relic/g`

---

Reviewed 1 of 23 files at r27, 5 of 5 files at r28, 1 of 1 files at r29, 1 of 1 files at r30, 1 of 1 files at r31, 2 of 2 files at r32, 2 of 2 files at r33, 1 of 1 files at r34, 18 of 19 files at r35, 1 of 4 files at r36, 5 of 5 files at r37, 1 of 1 files at r38.
Review status: 20 of 23 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/internal/client/client_test.go, line 1112 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQj-qb5VL0-x-h1rim:-LEQj-qb5VL0-x-h1rin:buk9knq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4058b2a45cb3f51910f682e45baacc1dc49e154/pkg/internal/client/client_test.go#L1112)):*
> ```Go
> 	// Do a Get using a different ctx (not the canceled one), and check that it
> 	// didn't take too long - take that as proof that it was not blocked on
> 	// intents.
> ```

The transaction would take a long time if the intents were blocked because .... (something about that transaction abandoned timeout)

---

*[pkg/internal/client/txn.go, line 664 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQjRmp3aXOLEAu6ggZ:-LEQjRmp3aXOLEAu6gg_:b-5frivk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4058b2a45cb3f51910f682e45baacc1dc49e154/pkg/internal/client/txn.go#L664)):*
> ```Go
> 
> 	ctx = txn.db.AnnotateCtx(context.Background())
> 	// !!! This needs to be a stopper task; plumb the stopper to the DB.
> ```

You're not merging this, right?

EDIT: squash the commit that fixes this.

---

*[pkg/internal/client/txn.go, line 934 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQk5r64tNU76cj5Apo:-LEQk5r64tNU76cj5App:b-51i3pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4058b2a45cb3f51910f682e45baacc1dc49e154/pkg/internal/client/txn.go#L934)):*
> ```Go
> 		sender = txn.mu.sender
> 		if txn.mu.Proto.Status != roachpb.PENDING || txn.mu.finalized {
> 			onlyRollback := lastIndex == 0 && haveEndTxn && !endTxnRequest.Commit
> ```

What's the point of allowing this rollback through when `txn.mu.Proto.Status != roachpb.PENDING`?

---

*[pkg/sql/conn_executor_prepare.go, line 225 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQiEhGEDA8LZ0kblOp:-LEQiEhH0m5ucAIlCYQn:b-45nvay) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa4d2babb1cfc0188fe7bcdd9ca7f9e73963c1d6/pkg/sql/conn_executor_prepare.go#L225)):*
> ```Go
> 		return nil, err
> 	}
> 	if err := txn.CommitOrCleanup(ctx); err != nil {
> ```

Should we commit this or roll it back? I agree that needing a transaction during stmt preparing is strange. It's even stranger that the txn would commit. If we're relying on this commit then that seems like a problem.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LEQijjm3jbE9O7RVhn2:b-tbr5kg)*

**Comment by andreimatei** [comment]
Review status: 20 of 23 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/client_test.go, line 1112 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQj-qb5VL0-x-h1rim:-LER252j61DByzEoC2wx:b-lkuq4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4058b2a45cb3f51910f682e45baacc1dc49e154/pkg/internal/client/client_test.go#L1112)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The transaction would take a long time if the intents were blocked because .... (something about that transaction abandoned timeout)
</blockquote></details>

added more words

---

*[pkg/internal/client/txn.go, line 664 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQjRmp3aXOLEAu6ggZ:-LER0l92-qZI8xF3_ets:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4058b2a45cb3f51910f682e45baacc1dc49e154/pkg/internal/client/txn.go#L664)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You're not merging this, right?

EDIT: squash the commit that fixes this.
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 934 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQk5r64tNU76cj5Apo:-LER2KXDERgQH6CvV9Ah:b-guu103) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4058b2a45cb3f51910f682e45baacc1dc49e154/pkg/internal/client/txn.go#L934)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's the point of allowing this rollback through when `txn.mu.Proto.Status != roachpb.PENDING`?
</blockquote></details>

well if the status is `Aborted`, there might still be intents to cleanup.
You might ask about allowing this when `txn.mu.finalized` is set and then I don't think I'd have an answer. But I'd leave this as is...

---

*[pkg/sql/conn_executor_prepare.go, line 225 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-LEQiEhGEDA8LZ0kblOp:-LER1730ACw6YJ7ft03z:b-4fj0n1) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa4d2babb1cfc0188fe7bcdd9ca7f9e73963c1d6/pkg/sql/conn_executor_prepare.go#L225)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we commit this or roll it back? I agree that needing a transaction during stmt preparing is strange. It's even stranger that the txn would commit. If we're relying on this commit then that seems like a problem.
</blockquote></details>

I agree that it's all crap, but I don't think that rolling back would be any better. If anything, trying to commit and getting an error might be an indication that the planning that was done... cannot be relied upon

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055)*

**Comment by andreimatei** [comment]
bors r+

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/23055#-:-LEkm-jC2CbYpJSCmIk_:b-ydi9ay)*

**Comment by tschottdorf** [comment]
Nice, glad to see this land!

**Pull request merged**